src/docfx/publish/PublishModelBuilder.cs

```diff
@@ -23,18 +24,25 @@ internal class PublishModelBuilder
         MonikerProvider monikerProvider,
         BuildOptions buildOptions,
         SourceMap sourceMap,
-        DocumentProvider documentProvider)
+        DocumentProvider documentProvider,
+        ContributionProvider contributionProvider)
     {
         _config = config;
         _errors = errors;
         _monikerProvider = monikerProvider;
         _locale = buildOptions.Locale;
         _sourceMap = sourceMap;
         _documentProvider = documentProvider;
+        _contributionProvider = contributionProvider;
     }
 
     public void AddOrUpdate(FilePath file, JObject? metadata, string? outputPath)
     {
+        if (metadata != null && file.Origin != FileOrigin.Redirection)
+        {
+            (_, var originalContentGitUrl, _) = _contributionProvider.GetGitUrl(file);
+            metadata.Add("source_url", originalContentGitUrl);
```

Comment:
Still, `source_url` is not a metadata, set it after line 64 of this file just like `SourcePath`. We don't need to touch `TocBuilder|ResourceBuilder|RedirectionBuilder`.

Reply:
got your point, just refine the code